object-cache.php

```diff
@@ -1000,46 +1044,60 @@ protected function _connect_redis() {
 			}
 		}
 
-		$this->redis = new Redis;
-
 		if ( file_exists( $redis_server['host'] ) && 'socket' === filetype( $redis_server['host'] ) ) { //unix socket connection
 			//port must be null or socket won't connect
 			$port = null;
 		} else { //tcp connection
 			$port = ! empty( $redis_server['port'] ) ? $redis_server['port'] : 6379;
 		}
-		$this->redis->connect( $redis_server['host'], $port, 1, null, 100 ); // 1s timeout, 100ms delay between reconnections
-		$keys_methods = array(
-			'auth'     => 'auth',
-			'database' => 'select',
+
+		$defaults = array(
+			'host' => $redis_server['host'],
+			'port' => $port,
+			'timeout' => 1000, // TODO I multiplied this by 1000 so we'd have a common measure of ms instead of s and ms, need to make sure this gets divided by 1000
+			'retry_interval' => 100,
 		);
+		// 1s timeout, 100ms delay between reconnections
+
+		// merging the defaults with the original $redis_server enables any
+		// custom parameters to get sent downstream to the redis client.
+		return array_replace_recursive( $redis_server, $defaults );
+	}
+
+	public function client_connection( $connection_details ) {
+		$redis = new Redis;
+
+		// TODO should we cache this connection?
+		$redis->connect(
+			$connection_details['host'],
+			$connection_details['port'],
+			// $connection_details['timeout'] is sent in milliseconds,
+			// connect() takes seconds, so divide by 1000
+			$connection_details['timeout'] / 1000,
+			null,
+			$connection_details['retry_interval']
+		);
+
+		return $redis;
+	}
+
+	function setup_client_connection( $redis, $settings, $keys_methods ) {
```

Comment:
Can you declare this a public method, and add PHPDoc to each new method you're defining?